### DISCUSSION TITLE
scheduler behaves weird

### DISCUSSION THREAD

**Discussion by matiasmicen** [apache/airflow#36220] [General]
Hi guys, I have a scheduled DAG that executes twice a day (10:00 and 18:00). The issue is that the 10:00 execution time is associated with a RUN ID corresponding to the logical date of 18:00. Therefore, the RUN ID name for the execution that occurs on 2023-12-14 at 10:00 is 'scheduled__2023-12-13T18:00:00+00:00'. In essence, I am executing the run from 12-13 one day later at 10:00.

This is my schedule interval: "00 10,18 * * 1-7"

So, how can I ensure that the logical date 'scheduled__2023-12-13T18:00:00+00:00' is executed on 2023-12-13 at 18:00?

<img width="1397" alt="image" src="https://github.com/apache/airflow/assets/146446960/be696d3d-b7e8-4e69-b461-73e7c8b39190">

Thanks.

**Comment by Taragolis** [comment]
This is intended behaviour for a long time, see [Scheduler](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/scheduler.html) documentation:

> [!TIP]
> If you run a DAG on a `schedule` of one day, the run with data interval starting on `2019-11-21` triggers after `2019-11-21T23:59`.
>
> **Let’s Repeat That**, the scheduler runs your job one `schedule` AFTER the start date, at the END of the interval.
> 
> You should refer to [DAG Runs](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/dag-run.html) for details on scheduling a DAG.

And this more reliable behaviour in cases when you work with batch processing, because in the case usual you wanted to for the already completed period: day, hour, week and etc

However not in all cases this might be suitable solution, as result in Airflow 2.2 [Timetables](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/timetable.html#timetables) was introduced as part of [AIP-39: Richer scheduler_interval](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-39+Richer+scheduler_interval)](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-39+Richer+scheduler_interval).

So your requirements to the scheduler decision could be achieved by use:
- Use builtin [`CronTriggerTimetable`](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/timetable.html#crontriggertimetable)
- Create custom timetable:
    - [How-to Guides: Customizing DAG Scheduling with Timetables](https://airflow.apache.org/docs/apache-airflow/stable/howto/timetable.html)
    - [Astronomer: DAG scheduling and timetables in Airflow](https://docs.astronomer.io/learn/scheduling-in-airflow#example-custom-timetable)

**Comment by matiasmicen** [comment]
Thanks Andrey, very much appreciated, I did it with Timetables.
